### PR TITLE
write_tsdb with metadata

### PR DESCRIPTION
**Pull request by ymettier**
Hello,

First of all, PR #1106 is a pre-requisite for this PR. This is why I based my work on that branch.

This patch allows use metadata to completely rewrite some metrics for OpenTSDB.
The problem is best explained in issue #709 and #887 : some metric names should not include the instance in their name but in a tag. For example, `cpu.1.cpu.user` should become `cpu.user` with an additional tag `cpu=1`

This patch is a proposition of a solution for this problem.

I added some metadata : 

<table>
<tr><th>metadata</th><th>explanation</th></tr>
<tr>
<td>tsdb_prefix</td><td>Will prefix the OpenTSDB &lt;metric&gt; (also prefix tsdb_id if defined)</td></tr>
<tr><td>tsdb_id</td><td>Replace the metric with this tag</td></tr>

<tr><td>tsdb_tag_plugin<br/>
tsdb_tag_pluginInstance<br/>
tsdb_tag_type<br/>
tsdb_tag_typeInstance<br/>
tsdb_tag_dsname</td>
<td>When defined, tsdb_tag_* removes the related item from metric id.<br />
If it is not empty, it will be the key of an opentsdb tag (the value is the item itself)<br />
If it is empty, no tag is defined.</td>
</tr></table>

Here is how I use them for commonly used plugins cpu, df, disk, interface, load and swap : 
```
<Chain "PreCache">
   <Rule "opentsdb_cpu">
     <Match "regex">
       Plugin "^cpu$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_pluginInstance" "cpu"
       MetaDataSet "tsdb_tag_type" ""
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
   <Rule "opentsdb_df">
     <Match "regex">
       Plugin "^df$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_pluginInstance" "mount"
       MetaDataSet "tsdb_tag_type" ""
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
   <Rule "opentsdb_disk">
     <Match "regex">
       Plugin "^disk$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_pluginInstance" "disk"
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
   <Rule "opentsdb_interface">
     <Match "regex">
       Plugin "^interface$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_pluginInstance" "iface"
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
   <Rule "opentsdb_load">
     <Match "regex">
       Plugin "^loac$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_type" ""
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
   <Rule "opentsdb_swap">
     <Match "regex">
       Plugin "^swap$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
 </Chain>
```

Of course, when you do not define metadata, it works just as before.

Regards,
Yves

### CODE REVIEW

**Comment by ymettier** [comment]
Hello,

Today's patch 32ce87f adds a new metadata tag : 
<table>
<tr><th>metadata</th><th>explanation</th></tr>
<tr>
<td>tsdb_tag</td><td>Add a literal "tagk=tagv" tag</td></tr>
</table>

This is useful for cases when pluginInstance (or typeInstance) may be set or not.

Example :  
```
LoadPlugin processes
<Plugin processes>
  ProcessMatch "collectd-agent" ".*/sbin/collectd.*collectd.conf.*-f.*"
</Plugin>

<Chain "OpenTSDB">
  <Rule "opentsdb_processes_one">
    <Match "regex">
      Plugin "^processes$"
      PluginInstance "^."
    </Match>
    <Target "set">
      MetaDataSet "tsdb_tag_pluginInstance" "process"
    </Target>
  </Rule>
  <Rule "opentsdb_processes_all">
    <Match "regex">
      Plugin "^processes$"
      PluginInstance "^$"
    </Match>
    <Target "set">
      MetaDataSet "tsdb_tag" "process=all"
    </Target>
  </Rule>
</Chain>
```

In this example, when the `pluginInstance` is set (e.g. when matching the collectd-agent process), the `pluginInstance` will be removed and we will have a tag `process=collectd-agent` (rule `opentsdb_processes_one`).
However, the processes plugin also monitor all processes information and sends metrics with `pluginInstance` not set. In this case (rule `opentsdb_processes_all`), I also add a tag named `process` with value `all`.

Here are some opentsdb `put` lines with this configuration : 
```
put processes.ps_state.running 1436276102 0 fqdn=myhost.mydomain process=all
put processes.ps_state.sleeping 1436276102 134 fqdn=myhost.mydomain process=all
put processes.ps_state.zombies 1436276102 0 fqdn=myhost.mydomain process=all

put processes.ps_count.threads 1436276102 9 fqdn=myhost.mydomain process=collectd-agent
put processes.ps_count.processes 1436276102 1 fqdn=myhost.mydomain process=collectd-agent
put processes.ps_cputime.user 1436276102 160000 fqdn=myhost.mydomain process=collectd-agent
put processes.ps_cputime.syst 1436276102 310000 fqdn=myhost.mydomain process=collectd-agent
```

Regards,
Yves

**Comment by nathanielc** [comment]
I have compiled and verified that this PR works for me. It works well to convert information from the identifier to tags. I used the config below and I was able to get data from collectd to InfluxDB with host specific tags and good metric names to fit the InfluxDB tagging scheme. I think many others will find this to be useful as it was exactly what I expected to be able to do out of the box.

```
LoadPlugin "write_tsdb"
LoadPlugin "match_regex"
LoadPlugin "target_set"

<Plugin write_tsdb>
  <Node>
    Host           "influxdb"
    Port           "4242"
    HostTags       "status=production deviceclass=www role=logging"
    StoreRates     false
    AlwaysAppendDS false
  </Node>
</Plugin>

<Chain "PreCache">
   <Rule "opentsdb_tags">
     <Match "regex">
       Plugin ".*"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_pluginInstance" "instance"
       MetaDataSet "tsdb_tag_type" "type"
       MetaDataSet "tsdb_tag_typeInstance" "type_instance"
       MetaDataSet "tsdb_tag_dsname" "dsname"
     </Target>
   </Rule>
</Chain>
```

**Comment by nathanielc** [comment]
@ymettier  I am getting tags without values with the above config for certain plugins. For example:

`put disk 1437510759 787985 fqdn=collect1 instance=vda type=disk_io_time type_instance= dsname=io_time`

Any idea why the tag key is being inserted without the value? Thanks

**Comment by nathanielc** [comment]
@ymettier  I changed the code locally to check for an empty value as well as an empty key:

In the define TSDB_STRING_APPEND_SPRINTF I added/modified these lines:

```c
const char *v = (value); \
if(k[0] != '\0' && v[0] != '\0') { \
```

This works for me and now tags are only added if they have a value.

**Comment by ymettier** [comment]
Thanks @nathanielc : I updated my code with your fix. See commit 87e08a2.

**Comment by ymettier** [comment]
Hello,

InfluxDB >= 0.9 users, please read !

As described (shortly) in https://influxdb.com/docs/v0.9/write_protocols/opentsdb.html, InfluxDB accepts metrics encoded for OpenTSDB.

So instead of sending "raw" Collectd metrics with the Collectd binary protocol, you can send them with write_tsdb. With this PR, you can easily rewrite the metrics names and add tags, exactly the same as it was planned for OpenTSDB.

In other words, write_tsdb and this PR work very well with InfluxDB >= 0.9 !!!

Regards,
Yves

**Comment by nathanielc** [comment]
+1 I have been running collectd patched with this PR and sending metrics to InfluxDB for a few weeks now. With the config I mentioned above I get a measurement per plugin and the rest of the metadata becomes tags. Its been really useful.

**Comment by dlmarion** [comment]
Running into this issue with collectd -> opentsdb. Thanks for working on this. I'm going to pull it down and give it a try.

**Comment by aferrari-technisys** [comment]
+1, some idea when will be merged into Master? this be amazing feature. Thanks to all that works around this.

**Comment by szop85** [comment]
Great job! Using it with colltecd/Graphite-API/Grafana/InfluxDB.

Cheers,
Szop

**Comment by szop85** [comment]
Hey guys,

I've run into a problem loading my configuration. I've split it up in multiple files and now I'm getting a:

```
 * Restarting statistics collection and monitoring daemon collectd                                                                                                                                                                                                                                                                                                          Found a configuration for the `network' plugin, but the plugin isn't loaded or didn't register a configuration callback.
Target `set': The `MetaDataSet' configuration option is not understood and will be ignored.
Target `set': The `MetaDataSet' configuration option is not understood and will be ignored.
Target `set': The `MetaDataSet' configuration option is not understood and will be ignored.
Target `set': The `MetaDataSet' configuration option is not understood and will be ignored.
Target `set': You need to set at least one of `Host', `Plugin', `PluginInstance' or `TypeInstance'.
Filter subsystem: Failed to create a set target.
```

Any idea what this means?

Cheers,
Szop

**Comment by cdgraff** [comment]
Hi All, I has the same error that reported before @szop85 

To check if I miss something, some can share his collect.conf file to use the same and see if I made some mistake?

The version I'm using be: collectd 5.5.0.377.g5afcfb3

Downloaded from:
https://ci.collectd.org/job/pull-requests-prepare-tarball/lastSuccessfulBuild/artifact/collectd-5.5.0.377.g5afcfb3.tar.gz

thanks for any advice!
Ale

**Comment by szop85** [comment]
Hey @cdgraff,

can you verify if you have the syslog plugin enabled?

Cheers,
Szop

**Comment by cdgraff** [comment]
Hi @szop85 no I don't has enabled, just in case i added and is the same error, just now this log into message file and I don't see in the console, but the message is the same.

I'm using the example provided by @nathanielc... if be working for you, can share in some Gist the config file? thanks!

**Comment by szop85** [comment]
After splitting my collectd config I forgot to activate/load the syslog plugin. After loading the syslog plugin, everything was fine. 

Cheers,
Szop

**Comment by ymettier** [comment]
```
Target `set': The `MetaDataSet' configuration option is not understood and will be ignored.
```
@szop85 : this is a feature of my patch. If you do not patch Collectd with PR #1107, Collectd cannot understand `MetaDataSet`.

Check if you successfully patched your Collectd and if you are really running the patched Collectd.

Regards,
Yves

**Comment by cdgraff** [comment]
@ymettier can show me a quick how to patch the master branch?

I do this and got and error:
[root@mon collectd]# git apply 1106.patch
[root@mon collectd]# git apply 1107.patch
1107.patch:795: trailing whitespace.

1107.patch:915: trailing whitespace.
                    while(NULL != (ptr2 = strchr(ptr2, ' '))) ptr2[0] = '_';
1107.patch:1022: trailing whitespace.

1107.patch:1040: trailing whitespace.

error: patch failed: src/collectd.conf.pod:8676
error: src/collectd.conf.pod: patch does not apply
error: patch failed: src/daemon/meta_data.c:106
error: src/daemon/meta_data.c: patch does not apply
error: patch failed: src/daemon/meta_data.h:43
error: src/daemon/meta_data.h: patch does not apply
error: patch failed: src/target_set.c:35
error: src/target_set.c: patch does not apply

I'm not good with git...
Thanks in advance!
Ale

**Comment by ymettier** [comment]
use "git merge" and merge my branch into Collectd master branch.
Then clean.sh, build.sh and the usual compile chain.

**Comment by matejzero** [comment]
I'm also testing this pull on my system for the last week and it is working very good. I would love to see this merged to master and in the next release, so I can move my system to InfluxDB

**Comment by szop85** [comment]
I'm having some trouble to query for the fqdn with InfluxDB. For example:

```
> select * from "users.users" limit 1
name: users.users
-----------------
time			deviceclass	fqdn		role	status		value
2015-09-15T14:25:59Z	www		dbm.local	logging	production	0
```
How would I do that?

Cheers,
Szop

**Comment by nathanielc** [comment]
@szop85 I am not sure what you are asking/expecting from that query... Try #influxdb on freenode or the influxdb mailing list and we can help you out.

**Comment by Dieken** [comment]
@kev009 and @octo,  any ETA this gets merged?  It's a very nice feature.

@nathanielc 's configuration works well for me, I don't have to configure MetaDataSet per plugin, that's very convenient.

```xml
LoadPlugin "match_regex"
LoadPlugin "target_set"

<Chain "PreCache">
    <Rule "opentsdb_tags">
        <Match "regex">
            Host        ""
        </Match>
        <Target "set">
            MetaDataSet "tsdb_prefix"               "sys."
            MetaDataSet "tsdb_tag_pluginInstance"   "instance"
            MetaDataSet "tsdb_tag_type"             "type"
            MetaDataSet "tsdb_tag_typeInstance"     "type_instance"
            MetaDataSet "tsdb_tag_dsname"           "dsname"
        </Target>
    </Rule>
</Chain>
```

**Comment by ymettier** [comment]
Hello,

Today's commit 1d584ae removes the `tsdb_tag` feature that I added earlier and replaces it with `tsdb_tag_add_XXXX`
<table>
<tr><th>metadata</th><th>explanation</th></tr>
<tr>
<td>tsdb_tag_add_<i>tagk</i></td><td>Add a literal "tagk=tagv" tag with tagk extracted from the tag name</td></tr>
</table>

This is useful for cases when pluginInstance (or typeInstance) may be set or not.

Example :  
```
LoadPlugin processes
<Plugin processes>
  ProcessMatch "collectd-agent" ".*/sbin/collectd.*collectd.conf.*-f.*"
</Plugin>

<Chain "OpenTSDB">
  <Rule "opentsdb_processes_one">
    <Match "regex">
      Plugin "^processes$"
      PluginInstance "^."
    </Match>
    <Target "set">
      MetaDataSet "tsdb_tag_pluginInstance" "process"
    </Target>
  </Rule>
  <Rule "opentsdb_processes_all">
    <Match "regex">
      Plugin "^processes$"
      PluginInstance "^$"
    </Match>
    <Target "set">
      MetaDataSet "tsdb_tag_add_process" "all"
    </Target>
  </Rule>
</Chain>
```

I'm really sorry to break the API I introduced earlier, but since this patch is not accepted yet, it's still time to do it.

The reason why I change this is because you cannot add more than one tsdb_tag. E.G. you can add as many Collectd metadata as you want, but the name of the Collectd metadata is unique. So there is only one tsdb_tag possible, which is not what we want.

The new idea is to add tags as metadata and prefix them with tsdb_tag_add_ to let the plugin write_tsdb know that it should add that tag.

Regards,
Yves

**Comment by ymettier** [comment]
Hello,

Today's commit improves logging when the connection to a [opentsdb/influxdb] becomes unavailable.
Instead of writing tons of messages in the log, you will get only one message. And one message when connection is back again.

Here is a sample of a log :
```
[2015-09-27 17:20:40] write_tsdb plugin: send failed with status -1 (Broken pipe)
[2015-09-27 17:20:40] write_tsdb plugin (localhost:4242): error with wt_send_message
[2015-09-27 17:20:40] write_tsdb plugin: Connecting to localhost:4242 failed. The last error was: Connection refused
[2015-09-27 17:29:00] write_tsdb plugin: Connecting to localhost:4242 succeeded.
```

Before that, I would have had the 3 first lines repeated every 10 secondes (interval = 10 for my tests) in the log.

Regards,
Yves

**Comment by ymettier** [comment]
Hello,

When TSDB is not available, write_tsdb will try to reconnect again and again. And perform DNS queries every time it tries to connect. That means 1 query for every metric for every interval. DNS flood !

This commt prevents such DNS flood with a new configuration option DNS_Cache_TTL that you may want to set to 60 (1 minute) or even more.

Regards,
Yves

**Comment by ymettier** [comment]
Hello,

My previous commit about DNS was not enough. Reason : there was still a DNS flood when the connection was lost (when you stop the tsdb server).

This commit improves the situation.
* because you don't want a DNS flood every `DNS_Cache_TTL` seconds, there is an additionnal random TTL (you can set it to 0 id you don't like it). Set it with `DNS_Random_Cache_TTL`.

* when the socket is closed, there is no DNS lookup. It will only try to reconnect with the latest DNS info. In most cases, that's enough. DNS lookup will happen after `DNS_Cache_TTL` + random(`DNS_Random_Cache_TTL`) seconds. That gives time to reconnect without the need for a refresh of the DNS info.

Default value for `DNS_Cache_TTL` has changed from 1 min to 10 min (check your DNS TTL to see if this is relevant)

Default value for `DNS_Random_Cache_TTL` is 15 x the plugin interval, e.g. 15 minutes if you set the interval to 60s. If you want to set it to a lower value, don't forget that Collectd will not send anything between 2 collect intervals. E.G. setting `DNS_Random_Cache_TTL` to only 1 x interval (or even less) is useless.

Regards,
Yves

**Comment by Dieken** [comment]
ping...any ETA this get merged? This patch works very well for me, thank you @ymettier !

**Comment by ntirupattur** [comment]
Bump -- Any ETA on this? I have tested this patch and works for me.

**Comment by ymettier** [comment]
Hello @octo (or any other maintainer).

This branch has now a conflict. It comes from commit 2d8020ff79f43f9abce82ef27b13b382c85e8f98 that removes a `close(cb->sock_fd);` line that I had also removed earlier in my branch.

How should I update my branch and PR ? `git merge master` (and mess the PR) ? `git pull --rebase upstream master` (and rewrite the history of my branch; mess the PR) ? Would you prefer a new PR ?

By the way, is there a reason why you don't merge this PR into master, or is it only a question of time ?

Regards,
Yves

**Comment by ymettier** [comment]
Hello again.

For your information, this PR fixes 2 issues I should have mentionned earlier instead of just fixing them in my branch :
* Issue #1148 (*write_tsdb plugin : DDOS attack on DNS*) fixed in a58cc3e51743cd92f6d8a0bb37a88394b588da8e and d3681238304204a90ed40de06e4d55b6bc098172
* Issue #1149 (*write_tsdb plugin : too many logs when TSDB is unavailable (/var full)*) fixed in b5b568e1ad23b39f12e1b504f8a8488eb6563ad7

Regards,
Yves

**Comment by kev009** [comment]
Functionally this LGTM

**Comment by Dieken** [comment]
when will this get merged? it's boring to patch and build collectd myself :-(

File: src/write_tsdb.c [hunk 1]
```diff
@@ -341,15 +436,106 @@ static int wt_format_values(char *ret, size_t ret_len,
...
+                while(NULL != (ptr2 = strchr(ptr2, ' '))) ptr2[0] = '_';  \
+                ptr += n; \
+                remaining_len -= n; \
+            } \
+        } \
+    } while(0)
+
+    if (vl->meta) {
+        TSDB_META_DATA_GET_STRING(meta_tag[TSDB_TAG_PLUGIN]);
+        if(temp) {
```
Comment (rubenk) [inline src/write_tsdb.c]: Space after if

File: src/write_tsdb.c [hunk 2]
```diff
@@ -341,15 +436,106 @@ static int wt_format_values(char *ret, size_t ret_len,
...
+            TSDB_STRING_APPEND_SPRINTF(temp, vl->plugin);
+            sfree(temp);
+        }
+
+        TSDB_META_DATA_GET_STRING(meta_tag[TSDB_TAG_PLUGININSTANCE]);
+        if(temp) {
```
Comment (rubenk) [inline src/write_tsdb.c]: Same here

File: src/write_tsdb.c [hunk 3]
```diff
@@ -341,15 +436,106 @@ static int wt_format_values(char *ret, size_t ret_len,
...
+            TSDB_STRING_APPEND_SPRINTF(temp, vl->plugin_instance);
+            sfree(temp);
+        }
+
+        TSDB_META_DATA_GET_STRING(meta_tag[TSDB_TAG_TYPE]);
+        if(temp) {
```
Comment (rubenk) [inline src/write_tsdb.c]: and here

File: src/write_tsdb.c [hunk 4]
```diff
@@ -341,15 +436,106 @@ static int wt_format_values(char *ret, size_t ret_len,
...
+            TSDB_STRING_APPEND_SPRINTF(temp, vl->type);
+            sfree(temp);
+        }
+
+        TSDB_META_DATA_GET_STRING(meta_tag[TSDB_TAG_TYPEINSTANCE]);
+        if(temp) {
```
Comment (rubenk) [inline src/write_tsdb.c]: and here

File: src/write_tsdb.c [hunk 5]
```diff
@@ -341,15 +436,106 @@ static int wt_format_values(char *ret, size_t ret_len,
...
+            TSDB_STRING_APPEND_SPRINTF(temp, vl->type_instance);
+            sfree(temp);
+        }
+
+        if(ds_name) {
```
Comment (rubenk) [inline src/write_tsdb.c]: and here

File: src/write_tsdb.c [hunk 6]
```diff
@@ -341,15 +436,106 @@ static int wt_format_values(char *ret, size_t ret_len,
...
+            TSDB_META_DATA_GET_STRING(meta_tag[TSDB_TAG_DSNAME]);
+            if(temp) {
```
Comment (rubenk) [inline src/write_tsdb.c]: here too

File: src/write_tsdb.c [hunk 7]
```diff
@@ -361,54 +547,92 @@ static int wt_format_name(char *ret, int ret_len,
...
-                ssnprintf(ret, ret_len, "%s%s.%s.%s", prefix, vl->plugin,
-                        vl->type_instance, vl->type);
-            }
-        } else { /* vl->plugin_instance != "" */
-            if (vl->type_instance[0] == '\0') {
-                ssnprintf(ret, ret_len, "%s%s.%s.%s", prefix, vl->plugin,
-                        vl->plugin_instance, vl->type);
+
+        for(i=0; i < (sizeof(meta_tag)/sizeof(*meta_tag)); i++) {
+            if(0 == meta_data_exists(vl->meta, meta_tag[i])) {
```
Comment (rubenk) [inline src/write_tsdb.c]: No Yoda conditionals please ;)

File: src/write_tsdb.c [hunk 8]
```diff
@@ -361,54 +547,92 @@ static int wt_format_name(char *ret, int ret_len,
...
+                /* defaults to already initialized format */
             } else {
-                ssnprintf(ret, ret_len, "%s%s.%s.%s.%s", prefix, vl->plugin,
-                        vl->plugin_instance, vl->type, vl->type_instance);
+                include_in_id[i] = 0;
             }
         }
     }
+    if(tsdb_id) {
```
Comment (rubenk) [inline src/write_tsdb.c]: Space after if, brace on the next line.

File: src/write_tsdb.c [hunk 9]
```diff
@@ -361,54 +547,92 @@ static int wt_format_name(char *ret, int ret_len,
...
+        ssnprintf(ret, ret_len, "%s%s", prefix?prefix:"", tsdb_id);
+    } else {
+#define TSDB_STRING_APPEND_STRING(string) do { \
+    const char *str = (string); \
+    size_t len = strlen(str); \
+    if(len > (remaining_len - 1)) { \
```
Comment (rubenk) [inline src/write_tsdb.c]: here too

File: src/write_tsdb.c [hunk 10]
```diff
@@ -361,54 +547,92 @@ static int wt_format_name(char *ret, int ret_len,
...
+        ptr[0] = '\0'; \
+        return(-ENOSPC); \
+    } \
+    if(len > 0) {  \
```
Comment (rubenk) [inline src/write_tsdb.c]: here too

File: src/write_tsdb.c [hunk 11]
```diff
@@ -361,54 +547,92 @@ static int wt_format_name(char *ret, int ret_len,
...
+        memcpy(ptr, str, len); \
+        ptr += len; \
+        remaining_len -= len; \
+    } \
+} while(0)
+
+#define TSDB_STRING_APPEND_DOT do { \
+    if(remaining_len > 2) {  \
```
Comment (rubenk) [inline src/write_tsdb.c]: here too

File: src/write_tsdb.c [hunk 12]
```diff
@@ -361,54 +547,92 @@ static int wt_format_name(char *ret, int ret_len,
...
+        return(-ENOSPC); \
+    } \
+} while(0)
+
+        char *ptr = ret;
+        size_t remaining_len = ret_len;
+        if(prefix) {
+            TSDB_STRING_APPEND_STRING(prefix);
+        }
+        if(include_in_id[TSDB_TAG_PLUGIN]) {
```
Comment (rubenk) [inline src/write_tsdb.c]: well, you get the idea.

File: src/collectd.conf.pod [hunk 13]
```diff
@@ -7165,7 +7167,34 @@ Synopsis:
...
+Global directives are:
+
+=over 4
+
+=item B<DNS_Cache_TTL> I<ttl>
+
+=item B<DNS_Random_Cache_TTL> I<ttl>
+
+When Collectd connects to a TSDB node, it will request the DNS. This can become
+a problem is the TSDN node is unavailable or badly configured because Collected
```
Comment (rubenk) [inline src/collectd.conf.pod]: s/Collected/Collectd/

**Comment by kev009** [comment]
Review fixes in https://github.com/collectd/collectd/pull/1655

**Comment by danoyoung** [comment]
Any ETA on this? We would like to see this feature as well

**Comment by octo** [comment]
Superseeded by #1655

**Comment by wpengcheng** [comment]
@ymettier My /etc/collectd.conf
LoadPlugin interface
<Plugin interface>
        Interface "eth0"
        IgnoreSelected false
</Plugin>
LoadPlugin match_regex
LoadPlugin target_set
<Chain "PreCache">
 <Rule "opentsdb_interface">
     <Match "regex">
       Plugin "^interface$"
     </Match>
     <Target "set">
       MetaDataSet "tsdb_tag_pluginInstance" "iface"
       MetaDataSet "tsdb_prefix" "sys."
     </Target>
   </Rule>
</Chain>

And systemctl restart collectd
tail -f /var/log/collectd.log
image

thank you very much！
